src/cs3iface.py

```diff
@@ -9,137 +9,215 @@
 import time
-import os
+import http
+import requests
 import grpc
 
-import cs3.gateway.v1beta1.gateway_api_pb2 as cs3gw
+#from tusclient import client as tusclient
+
+#from google.auth.transport import grpc as google_auth_transport_grpc
+#from google.auth import jwt as google_auth_jwt
+#from google import auth as google_auth
+
+import cs3.storage.provider.v1beta1.resources_pb2 as spr
+import cs3.storage.provider.v1beta1.provider_api_pb2 as sp
 import cs3.gateway.v1beta1.gateway_api_pb2_grpc as cs3gw_grpc
+import cs3.gateway.v1beta1.gateway_api_pb2 as cs3gw
+import cs3.rpc.code_pb2 as cs3code
+
 
 # module-wide state
-config = None
-log = None
-tok = None
+ctx = {}        # "map" to store some module context: cf. init()
+tokens = {}     # map userid [string] to authentication token
 
-def _getfilename(filename):
-  '''map the given filename into the target namespace by prepending the homepath (see storagehomepath in wopiserver.conf)'''
-  return filename   # TODO do we need that??
 
-def init(inconfig, inlog):
+def init(config, log):
   '''Init module-level variables'''
-  global config         # pylint: disable=global-statement
-  global log            # pylint: disable=global-statement
-  global tok            # pylint: disable=global-statement
-  config = inconfig
-  log = inlog
-  #revaurl = config.get('cs3', 'revaurl')
-  revaurl = 'iota:19000' # XXX for now
-
+  ctx['log'] = log
+  ctx['chunksize'] = config.getint('io', 'chunksize')
+  revahost = config.get('cs3', 'revahost')
   # prepare the gRPC connection
-  ch = grpc.insecure_channel(revaurl)
-  cs3stub = cs3gw_grpc.GatewayAPIStub(ch)
-  authReq = cs3gw.AuthenticateRequest(type='basic', client_id='gonzalhu', client_secret='test')
-  tok = cs3stub.Authenticate(authReq)
-
-
-def stat(_endpoint, filename, _ruid, _rgid):
-  '''Stat a file and returns (size, mtime) as well as other extended info. Assume the given uid, gid has access.'''
-  filename = _getfilename(filename)
-  try:
-    tstart = time.clock()
-    statInfo = os.stat(filename)
-    tend = time.clock()
-    log.info('msg="Invoked stat" filename="%s" elapsedTimems="%.1f"' % (filename, (tend-tstart)*1000))
+  ch = grpc.insecure_channel(revahost)
+  ctx['cs3stub'] = cs3gw_grpc.GatewayAPIStub(ch)
+
+
+def _authenticate(userid):
+  '''Obtain a token from Reva for the given userid'''
+  # TODO this will become
+  #authReq = cs3gw.AuthenticateRequest(type='bearer', client_secret=userid)
+  authReq = cs3gw.AuthenticateRequest(type='basic', client_id='einstein', client_secret='relativity')
+  if userid not in tokens:
+    # authenticate this user
+    authRes = ctx['cs3stub'].Authenticate(authReq)
+    tokens[userid] = authRes.token
+  return tokens[userid]
+
+
+def stat(endpoint, fileid, userid):
+  '''Stat a file and returns (size, mtime) as well as other extended info using the given userid as access token.'''
+  if endpoint == 'default':
+    raise IOError('A CS3API-compatible storage endpoint must be identified by a storage UUID')
```

Comment:
I'm slightly unsure what's included in `endpoint` - maybe this can be documented (if we cannot rename it)? (If it's not an endpoint but rather a storage id?)

Reply:
Indeed `endpoint` == storage id, let's document it in the comment but I'd prefer not to change the signature (even though it's possible in python) to keep it consistent with the other interfaces.

Reply:
✅ What does endpoint contain in the other cases?

Reply:
for local storage it's ignored, for xroot it's the eos instance like `root://eoshome-a`

Reply:
Can we call it something like "storageindetifier" then?

Reply:
For xroot it's really an endpoint, a target URL, not just an id. I still prefer `endpoint` (also shorter than `storageidentifier`).

Reply:
I'm not convinced but let's keep it like this for now.